avocado/utils/kernel.py

```diff
@@ -44,15 +45,18 @@ def __init__(self, version, config_path=None, work_dir=None,
         :param data_dirs: list of directories to keep the downloaded kernel
         :return: None.
         """
+        self.URL = src_url
         self.version = version
         self.config_path = config_path
+        self.distro = distro.detect()
         if work_dir is None:
             work_dir = tempfile.mkdtemp(prefix='avocado_' + __name__)
         self.work_dir = work_dir
         if data_dirs is not None:
             self.data_dirs = data_dirs
         else:
             self.data_dirs = [self.work_dir]
+        self.do_install=do_install
```

Comment:
As per my previous comment, I believe there's no need to set this as a class parameter, and then as an attribute.  But, here goes a style tip: when doing other assignments, please add a space separating the LHS, equal sign, and RHS.

Reply:
Should we turn this as method parameter ?

Reply:
I don't understand how this can be added as a method (which one?) parameter.  IMO this fits into `install()`, but it wouldn't need any extra parameter.

Reply:
sorry for lack of details in my previous question.
I meant to ask whether the do_install flag can be implemented as an argument of the method build(). ei:

```diff
diff --git a/avocado/utils/kernel.py b/avocado/utils/kernel.py
index 37b41d2f..4e93de7d 100644
--- a/avocado/utils/kernel.py
+++ b/avocado/utils/kernel.py
@@ -100,13 +100,17 @@ class KernelBuild(object):
             dotconfig = os.path.join(self.linux_dir, '.config')
             shutil.copy(self.config_path, dotconfig)

-    def build(self):
+    def build(self, build_for_install=none):
         """
         Build kernel from source.
+
+        :param build_for_install: when not None, the appropriate
+                                  platform package is built
+                                  for install() to use
         """
         log.info("Starting build the kernel")
         build_output_format = ""
-        if self.do_install is not None:
+        if build_for_install is not None:
             if self.distro.name == "Ubuntu":
                 build_output_format = "deb-pkg"
         if self.config_path is None:
```

Reply:
There's nothing *wrong* with your proposal for a `build_for_install` parameter to `build()`, but when designing APIs, it makes sense to brainstorm a bit more than usual, so, let's do it.

Initially, based on my previous comment (https://github.com/avocado-framework/avocado/pull/2527#discussion_r175624169), I thought it made more sense to have a generic (unchanged) `build()` method, and the `install()` method would handle the last build steps too (assuming most of the build would already be done, and the packaging would take little time and could be absorbed by it).

But, now I think your idea is better.  The target to be built, be it a stock "vmlinuz" image or a package, is really *build* related, and not installation related.  So, having said that, I'd only tweak the type of the parameter (bool) and optionally the name (suggestion: `binary_package`).

Reply:
Please let me know if this change would work:

```diff
@@ -99,17 +99,18 @@ class KernelBuild(object):
             dotconfig = os.path.join(self.linux_dir, '.config')
             shutil.copy(self.config_path, dotconfig)

-    def build(self, build_for_install=None):
+    def build(self, binary_package=None):
         """
         Build kernel from source.

-        :param build_for_install: when not None, the appropriate
+        :param binary_package: when not None, the appropriate
                                   platform package is built
                                   for install() to use
+        :type binary_pacakge: bool or None
         """
         log.info("Starting build the kernel")
         build_output_format = ""
-        if build_for_install is not None:
+        if binary_package is not None:
             if self.distro.name == "Ubuntu":
                 build_output_format = "deb-pkg"
         if self.config_path is None:
```

Reply:
What I meant by using a bool is that `binary_package=False` can be the default. The parameter would simply be:

```
:type binary_pacakge: bool
```

No need for None here.

Reply:
did I get it right?

```diff
diff --git a/avocado/utils/kernel.py b/avocado/utils/kernel.py
index ee566f52..9e7db63c 100644
--- a/avocado/utils/kernel.py
+++ b/avocado/utils/kernel.py
@@ -99,18 +99,18 @@ class KernelBuild(object):
             dotconfig = os.path.join(self.linux_dir, '.config')
             shutil.copy(self.config_path, dotconfig)

-    def build(self, binary_package=None):
+    def build(self, binary_package=False):
         """
         Build kernel from source.

-        :param binary_package: when not None, the appropriate
+        :param binary_package: when True, the appropriate
                                   platform package is built
                                   for install() to use
-        :type binary_pacakge: bool or None
+        :type binary_pacakge: bool
         """
         log.info("Starting build the kernel")
         build_output_format = ""
-        if binary_package is not None:
+        if binary_package is True:
             if self.distro.name == "Ubuntu":
                 build_output_format = "deb-pkg"
         if self.config_path is None:
```